enhancements/cvo/upgrade_preflight_checks.md

```diff
@@ -0,0 +1,105 @@
+authors:
+  - "@LalatenduMohanty"
+reviewers:
+  - TBD
+approvers:
+  - TBD
+creation-date: 2020-06-04
+last-updated: 2020-06-05
+status: implementable
+---
+
+# Upgrade Preflight Checks For Operators
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions [optional]
+
+* How CVO can get the list of operators which has the preflight checks implemented?
+* What information CVO needs from the operators to create the preflight jobs?
+* Does CVO already has capability to create jobs for preflight checks?
+* How to trigger execution of preflight checks?
+* How CVO will report when it gets the result from preflight checks?
+* What information CVO needs from operators to create the preflight job?
+
+## Summary
+
+The goal of over the air update(OTA) feature of OpenShift is to make upgrades seemless and make the upgrade a success almost always.
+To improve the upgrade success we should be able to detect if the new version is compatible with the existing version before running the actual upgrade.
+As OpenShift V4 is composed of many operators, we need to give operators abiltity to run some checks to findout if the new version is compatible with the existing version and use the result to improve the successrate of upgrades.
+
+## Motivation
+
+CVO managed operators should be able to run preflight checks before the actual upgrade.
+It would help operators to findout if they can move to the specific version safely before doing an actual upgrade.
+Also administrators are empowered to run these preflight checks without doing an upgrade.
+
+
+### Goals
+
+* Administrators will be able to run preflight checks without running the upgrade.
+* Give Operators (CVO managed) ability to add checks to check if they can safely upgrade to a given release.
+* Preflight checks would be idempotent in nature.
+
+### Non-Goals
+
+* This enhancement will not add code required for preflight checks for individual operators.
+* Operators will have to figurout the actual code they would like to run as part of preflight checks.
+* The preflight checks should not be used as a migration mechanism.
+
+## Proposal
+
+Operators need to provide a different bin path for the preflight checks i.e. /bin/preflight.
+The binpath need to exist inside of the operator image.
+Cluster version operator (CVO) will be running a single job per operator for the preflight checks.
+The binpath /bin/preflight must be executable and it can be symbolic link to the operator binary or a script or It can be non existant.
+CVO will not fail if the path does not exist in an operator image but it would assume that preflight checks are not implemented for the respective operator.
```

Comment:
It seems non-trivial to expect the CVO to know the difference between my Job failed because the binary doesn't exist, or the job failed because my preflights failed. We wouldn't want the CVO having to extract/parse operator images checking if that exists. 

I'm curious if the operators should "register" that they have a preflight check, at which point if that path doesn't exist, your preflight check is bad.

Reply:
How do we identify operators?  For example:

```console
$ oc adm release info quay.io/openshift-release-dev/ocp-release:4.4.6-x86_64
Name:      4.4.6
...
  cli                                            sha256:b0edd74f803bc66809975af282244b1c2d9d6c72d90ac86c24f5f626be236079
...
  cloud-credential-operator                      sha256:cb8f27c8821b93ad32feeccce5991b28f09b62e19f8365275eb73db3c7d87acd
...
```

The referenced `cloud-credential-operator` image is an operator, while the referenced `cli` image is not.  If operators have to register (via a new manifest type?) as suggested by @dgoodwin, that would address this issue.  But, if operators have to register, then preflight checks would not cover "can my cluster pull all the referenced images?", which is something we want to cover in preflights (and which should probably get listed under *Goals*).  Perhaps the preflight logic could be:

1. For every image referenced by the target release image...
    1. Create a Job using the referenced image and running `core-update-preflight`.  The more-specific command makes it less likely that an image has a matching command with different semantics.  Dropping `/bin/` in favor of `PATH` avoids us having more opinions about the referenced images than necessary.
    2. Collects the output of the completed job.
    3. Counts both exit code 0 and 127 (command not found, [in POSIX][1]) as a successful preflight.

[1]: https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_08_02

Reply:
Feels like that could be a fairly large number of jobs launched to just fail. The experience would be a lot more responsive if we could just do it for operators opting in.

Reply:
> Feels like that could be a fairly large number of jobs launched to just fail.

"Job exits 127 (command not found)" is not "job failed".  It would be "we can successfully pull this important image, and the image did not define further preflight checks".  Or do you consider answering "can my cluster pull all the referenced images?" to be a non-goal?

Reply:
I'm more referring to the process of a "can I upgrade to version x" launching what, 40 odd jobs in the cluster, pulling 40 images, when a fraction of that actually needs to happen. It's a moderate chunk of work to schedule and would bloat the response time for an admin trying to do it interactively.

Reply:
I'm still thinking it's better to have an opt-in or registration mechanism, and we launch only jobs for work we know needs to be done.

Reply:
I think we're closer to a hundred images ;).  And if we were pulling through our local registry, that would be useful cache warming, not wasted traffic.  But, given that we aren't pulling through today, sounds like you're saying "non-goal".

Reply:
I'm in favor of pre-flighting all desired version operators. Verification that the image is available with our pull secret should be an explicit goal of this effort.

Reply:
We need the new images for the preflight checks as the code for preflight check with be in the new image.

Reply:
But I understand @dgoodwin 's comment about `We wouldn't want the CVO having to extract/parse operator images checking if that exists.` Not sure if we have alternative. Will get back on this.

Reply:
I don't think we _need_ separate images; operators could have a preflight command in an existing image.  We could build this to provide folks with a way to use preflight-specific images, but that would weaken the "can we pull (some) images that matter?" goal, so I don't see much point.

Reply:
New images = The new operator images from the release payload to which the upgrade need to happen.

Reply:
Even if we download and verify all the images associated with a release payload, we should not just starts jobs for all of the operators if preflight checks are not implemented. WDYT about having a new manifest whose purpose is only to tell CVO about the information required to start a job.

Reply:
I'm +1 for any mechanism by which an operator can indicate "I have pre-flight checks to run" that doesn't require pulling introspecting each image.

Reply:
Added a separate section for this under `Implementation Details` https://github.com/openshift/enhancements/commit/6f74b40a76776ac4a922bc3c22320c1d6d5f89ee